internal/events/block_confirmations.go

```diff
@@ -0,0 +1,520 @@
+	bcm.log.Debugf("Downloaded block header by hash: %d / %s parent=%s", blockInfo.Number, blockInfo.Hash, blockInfo.ParentHash)
+
+	bcm.addToCache(blockInfo)
+	return blockInfo, nil
+}
+
+func (bcm *blockConfirmationManager) getBlockByNumber(blockNumber uint64, expectedParentHash string) (*blockInfo, error) {
+	cached, ok := bcm.blockCache.Get(strconv.FormatUint(blockNumber, 10))
+	if ok {
+		blockInfo := cached.(*blockInfo)
+		parentHash := blockInfo.ParentHash.String()
+		if parentHash != expectedParentHash {
+			// Treat a missing block, or a mismatched block, both as a cache miss and query the node
+			bcm.log.Debugf("Block cache miss due to parent hash mismatch: %d / %s parent=%s required=%s ", blockInfo.Number, blockInfo.Hash, parentHash, expectedParentHash)
+		} else {
+			return blockInfo, nil
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(bcm.ctx, 30*time.Second)
+	defer cancel()
+	var blockInfo *blockInfo
+	err := bcm.rpc.CallContext(ctx, &blockInfo, "eth_getBlockByNumber", ethbinding.HexUint64(blockNumber), false /* only the txn hashes */)
+	if err != nil {
+		return nil, errors.Errorf(errors.RPCCallReturnedError, "eth_getBlockByNumber", err)
+	}
+	if blockInfo == nil {
+		return nil, nil
+	}
+	bcm.log.Debugf("Downloaded block header by number: %d / %s parent=%s", blockInfo.Number, blockInfo.Hash, blockInfo.ParentHash)
+
+	bcm.addToCache(blockInfo)
+	return blockInfo, nil
+}
+
+func (bcm *blockConfirmationManager) confirmationsListener() {
+	defer close(bcm.done)
+	pollTimer := time.NewTimer(0)
+	notifications := make([]*bcmNotification, 0)
+	for {
+		popped := false
+		for !popped {
+			select {
+			case <-pollTimer.C:
+				popped = true
+			case <-bcm.ctx.Done():
+				bcm.log.Debugf("Block confirmation listener stopping")
+				return
+			case notification := <-bcm.bcmNotifications:
+				if notification.nType == bcmStopStream {
+					// Handle stream notifications immediately
+					bcm.streamStopped(notification)
+				} else {
+					// Defer until after we've got new logs
+					notifications = append(notifications, notification)
+				}
+			}
+		}
+		pollTimer = time.NewTimer(bcm.pollingInterval)
+
+		// Setup a filter if we're missing one
+		if bcm.filterStale {
+			if err := bcm.createBlockFilter(); err != nil {
+				bcm.log.Errorf("Failed to create block filter: %s", err)
+				continue
+			}
+
+			if err := bcm.walkChain(); err != nil {
+				bcm.log.Errorf("Failed to create walk chain after restoring filter: %s", err)
+				continue
+			}
+		}
+
+		// Do the poll
+		blockHashes, err := bcm.pollBlockFilter()
+		if err != nil {
+			bcm.log.Errorf("Failed to retrieve blocks from filter: %s", err)
+			continue
+		}
+
+		// Process each new block
+		bcm.processBlockHashes(blockHashes)
+
+		// Process any new notifications - we do this at the end, so it can benefit
+		// from knowing the latest highestBlockSeen
+		if err := bcm.processNotifications(notifications); err != nil {
+			bcm.log.Errorf("Failed processing notifications: %s", err)
```

Comment:
```suggestion
			bcm.log.Errorf("Failed to process notifications: %s", err)
```
along the lines of other messages..